### ISSUE TITLE
Uploaded image isn't in an allowed aspect ratio :: 1.6.1

### ISSUE THREAD

**Issue by beshoo1** [mgp25/Instagram-API#1072]
@mgp25  I am optimizing the resize class that i sent  you before. but i need to understand what is the correct allowed aspect ratio that instagram likes .

since while i am trying to upload  640x320 return an error !!

```
POST:  media/configure/
DATA: ig_sig_key_version=4&signed_body=778e3f3d227a681b7b680b414e812d699621024a2e2eb8ef8ef1d5a6c8481464.{"_csrftoken":"OLHpWCvpUzYnuY1NIXNYhvs2oeSP5zS6","media_folder":"Instagram","source_type":4,"_uid":"3986477616","_uuid":"fcea4419-e1b1-434d-9cb4-e6592e1f62ec","caption":"","upload_id":"1489328979093","device":{"manufacturer":"Fly","model":"FS405","android_version":18,"android_release":"4.3"},"edits":{"crop_original_size":[640,640],"crop_center":[0,0],"crop_zoom":1},"extra":{"source_width":640,"source_height":640}}
← 400    80B
RESPONSE: {"status": "fail", "message": "Uploaded image isn't in an allowed aspect ratio"}



What is the best guideline  for this ?

**Comment by SteveJobzniak** [comment]
640x320 is ratio 2:1. Instagram allows 1:1 and SOME slightly wider aspects, but definitely not media that is TWICE as wide as it is tall.

Instagram places all media in 1:1 aspect ratio boxes (squares). Recently they added support for landscape and portrait (non-square) images.

The longest side is never allowed to anything other than these dimensions: `16x16`, `30x30`, `32x32`, `48x48`, `50x50`, `60x60`, `64x64`, `75x75`, `80x80`, `96x96`, `100x100`, `128x128`, `150x150`, `160x160`, `192x192`, `200x200`, `240x240`, `320x320`, `480x480`, `600x600`, `640x640`, `750x750`, `800x800`, `960x960`, `1080x1080`. There _may_ be more dimensions that I haven't discovered, but it's unlikely.

That means that the allowed 1:1 square aspect ratios are the EXACT dimensions above.

If an image does not exactly fit that size, then it must be resized and possibly also cropped.

Uneven aspects (portrait/landscape) are allowed to have one side shorter than that. But by how much? I don't know. Perhaps 1.6:1 is the max they allow. So 1:1 to 1.6:1.

Meaning that a 640x640 box image is allowed to be `floor(640/1.6)x640` (portrait), or `640xfloor(640/1.6)` (landscape). But to do this, you must analyze the input image, see what aspect ratio it is, center-crop its largest dimension (its width or its height) if it exceeds 1.6 ratio. And then resize everything downwards by snapping its BIGGEST dimension to the nearest LOWER box-size of all the ones that I listed above.

Seeing past comments from you, I think this work may be a bit more than you can handle, but I wish you luck.

**Comment by ping** [comment]
I'm quite doubtful that there is a fixed set of values for the longest side of a media.

Examples: 

Media ID: 1465582645818869466_329452045
1080x1349 http://scontent-ord1-1.cdninstagram.com/t51.2885-15/e35/17076450_1016274778503422_8886262098196168704_n.jpg

Media ID: 1465989456212471257_329452045
1080x1274 http://scontent-ord1-1.cdninstagram.com/t51.2885-15/e35/17076371_179887479181113_3896725043412992_n.jpg

In an unscientific sample survey of about 1000 IG posts that I did a few months ago, the accept range of media aspect ratio (width/height) is between 0.8 to 1.9. For stories, the range is 0.5625 to 0.75.

**Comment by SteveJobzniak** [comment]
@ping I am 100% sure that there is a fixed size. But I misspoke above. The fixed size is for the SHORTEST side of a media.

**Comment by ping** [comment]
@SteveJobzniak Found another example, shortest side 722

Media ID: 1458070254172474127_329452045
1080x722 http://scontent-ord1-1.cdninstagram.com/t51.2885-15/e35/16789052_718854271612223_7409841672579710976_n.jpg

**Comment by SteveJobzniak** [comment]
@ping Hmm.

That's interesting.

I tried 722x722:

![screen shot 2017-03-12 at 17 59 25](https://cloud.githubusercontent.com/assets/22373193/23833922/a74e0036-074d-11e7-882d-7ae11af77827.png)

http://scontent-ord1-1.cdninstagram.com/t51.2885-15/p722x722/e35/16789052_718854271612223_7409841672579710976_n.jpg

But the typical 640x640 works:

http://scontent-ord1-1.cdninstagram.com/t51.2885-15/p640x640/e35/16789052_718854271612223_7409841672579710976_n.jpg

So 722 isn't allowed by the "/pXxX" resize-API... but they actually allow other sizes for raw media files. That's odd.

Edit: Oh wait... your example is 1080x722...

So yeah, that means that the rule isn't about a particular side. It's about ONE side MUST be one of the sizes I listed. But it doesn't matter if the other side is shorter or longer than that! I'll update my comment again.

**Comment by ping** [comment]
No comments about the OP. :)

Media ID: 1439657673842063763_329452045
450x300 http://scontent-ord1-1.cdninstagram.com/t51.2885-15/e35/16228537_257600638012348_5528412032883228672_n.jpg

If you don't mind me asking, how about you getting this set of sizes? I'm quite annoyed that I misplaced my original dataset so I cannot more accurately validate your list. If I can find the time, I may run another sample survey again to see if the sizes/aspect ratios still match up.

**Comment by SteveJobzniak** [comment]
@ping Hehe you don't have to comment on OP :D I am just annoyed because all his attempts to hide himself from Instagram makes me think he's doing some bannable offense against Instagram (such as spamming, auto-follower bots, etc) and we don't allow that here.

> If you don't mind me asking, how about you getting this set of sizes?

With very tedious tests via their image resizing API to check for all valid box-sizes. I did it manually but I guess I could have written something that tried 1080 requests in 1-pixel increments.

Basically they generate all kinds of standard HD (1080), standard (640), icon (96, 64, 48, 32, 16), etc. So I tried about 50 logical sizes and found the ones above.

Just take the URLs I showed above, and edit their "/pXxX/" flag. The X and X MUST always be identical (a square box). And must be an allowed box-size.  Otherwise the server replies with 5xx Server Error.

**Comment by SteveJobzniak** [comment]
There's also a "/sXxX/" flag, by the way.

The s (square downsize) flag forces the maximum dimension of the image to fit within the provided value. In other words, it means “downsize so that it definitely fits completely within this square box”.

The p (portrait/landscape downsize) flag forces the minimum dimension (whether that’s the width or the height) to fit within the provided value. In other words, it means “downsize so that at least the width or the height of the image fits within this box, but let the other, larger side remain larger and stick out of the box if it needs to”.

**Comment by SteveJobzniak** [comment]
And if you're curious: Right now Instagram only officially uses resizing on their video thumbnails. The on-server thumbnail is generated in the same dimensions as the video file itself, and they use one of the two downsizing flags to scale it to any thumbnail size they need.

**Comment by ping** [comment]
@SteveJobzniak Ah, you're talking about the Instagram resizing dimensions. Yes, I'm aware of the /s /p sizing rules.

Although, I believe that OP (whatever his intentions may be) was talking about the allowable dimensions for *upload*, which does not, in my opinion, need to fall into any of the /s /p standard sizing. 

I take cue from the "original_height" / "original_width" media attributes returned by the api, which in my experience can vary quite significantly.

Another case in point, media id: 1435497474474607135_329452045
386x386 http://scontent-ord1-1.cdninstagram.com/t51.2885-15/e35/16124360_1391557434211763_308824851710214144_n.jpg

**Comment by SteveJobzniak** [comment]
@ping Is it possible that their server handles the resizing for us automatically? I thought the app did that before uploading the media.

**Comment by ping** [comment]
@SteveJobzniak In my experience, the app seems to resize the media to a reasonable size (which I suspect is max side of 1280 since I haven't seen any original_width/original_height values greater than that) and the server handles the rest.

**Comment by SteveJobzniak** [comment]
Ahh. That makes sense. So the client has to pre-resize to a reasonable size (instead of the huge raw file stored in the phone's library).

And the client seems to do cropping too, judging by beshoo's error message "Uploaded image isn't in an allowed aspect ratio".

That means the client should crop to an allowed ratio, and resize so that no size is > 1280. The rest is handled by the server.

All of these functions should probably be a separate function or library entirely. Let's not put it in the uploadTimelinePhoto/uploadStoryPhoto/uploadTimelineAlbum... that would quickly get super messy with extra parameters.

**Comment by SteveJobzniak** [comment]
All of these functions should probably be a separate function or library entirely. Let's not put it in the uploadTimelinePhoto/uploadStoryPhoto/uploadTimelineAlbum... that would quickly get super messy with extra parameters.

A class for resizing images before upload would definitely be useful. Then users could do something like:

```php
$resizer = \InstagramAPI\ImageResizer();
$tmpimage = $img->process($path, $cropOffset);
$i->uploadTimelinePhoto($tmpimage);
unlink($tmpimage);
```

The idea above being: process() crops the image to the max-allowed aspect ratio, and places the cropping either centered (if cropOffset is null), or at the percentage position in cropOffset. And then it resizes so that no side is larger than 1280. And finally does "imagejpg()" with 100% quality to create a temp result image in /tmp.

**Comment by SteveJobzniak** [comment]
I haven't got any time but all of that would be pretty simple and quite fun to code.

PHP's image* functions are well documented: http://php.net/manual/en/ref.image.php.

It would have to:

1. Detect the input file format somehow.
2. Analyze the aspect ratio.
3. Crop it automatically, and place the "focal point" of the crop at the center or at the cropOffset percentage offset from the center (expressed as a number like -20 for focusing at 20% before the center or 20 for 20% after the center).
4. Resize it to max 1280 in one dimension.
5. Write it to a temp-file file at 100% quality. Use one of the "generate temp filename" functions in PHP for that.
6. Destroy the image resources via imagedestroy() to free the PHP memory again.
7. Return the filename.
8. Tell the user to delete the image via unlink(). Alternatively, wrap the returned image in a class object with a __destructor() which unlinks the file automatically. That way the file will be deleted as soon as the object goes out of scope (such as when the user's personal "uploadwhatever()" function ends). And the class should have a separate ->delete() function too, to delete ahead of time if desired.

That would be nice and would be an evening of coding. Interested, ping? ;)

**Comment by SteveJobzniak** [comment]
```php

class ImageAutoResizer
{
    protected $_inputFile;
    protected $_cropOffset;
    protected $_outputFile;

    public function __construct($inputFile, $cropOffset = null)
    {
       if (!is_file($inputFile)) {
           throw new \InvalidArgumentException("Input file doesn't exist.");
       }
       $this->_inputFile = $inputFile;
       $this->_cropOffset = $cropOffset;
    }

    public function delete()
    {
        if (!is_null($this->_outputFile) && is_file($this->_outputFile)) {
            return unlink($this->_outputFile);
        }
        return true;
    }

    public function getFile()
    {
        if (is_null($this->_outputFile)) {
            $this->_process();
        }
        return $this->_outputFile;
    }

    protected function _process()
    {
        // TODO: process input, make tempfile...
        $this->_outputFile = $tmpFile;
    }
}
```

There... That's the shell for the class written here on Github haha. :)

**Comment by ping** [comment]
@SteveJobzniak It is probably important to note that standard timeline posts vs album posts vs story posts have different aspect ratios, so the crop ratio has to be user-defined.

I agree that the media manipulation should be done in a separate class. But I think the various upload* functions can do with some basic sanity checking for dimensions (e.g. a 3000x3000 upload just isn't gonna fly). If the aspect ratio ranges can be nailed down, that too can be checked in the upload* functions. For videos, min (3secs) max (60secs) duration can also be checked. If the checks fail, an appropriate exception should be thrown (maybe something like an InvalidMediaException).

My contributions here are largely to do with reverse-engineering, so I'll have to defer to your expertise in PHP. :) Great job so far by the way. 👍

**Comment by SteveJobzniak** [comment]
@ping Hmm. Well "basic sanity checking" would cost a lot of performance. It means loading the image and decoding it, then checking http://php.net/manual/en/function.imageresolution.php

So I think maybe users should get used to always using the ImageAutoResizer class.

And as for letting user control aspect ratio: Hmm okay, sounds nasty though... The user shouldn't have to know about them. I guess they can be set via constants. I'll update the class example above.

**Comment by SteveJobzniak** [comment]
Updated.

**Comment by SteveJobzniak** [comment]
Maybe I'll have energy for this one of the coming days. But I hope someone else finishes it if I don't. I have burned myself out for a while with the non-stop coding every waking minute for 10 days, haha.

**Comment by ping** [comment]
@SteveJobzniak Again, I defer to your judgement about the technical implementation and constraints. :) IMHO, this can be scheduled for implementation later, after v2 is stable / officially released. Take it easy mate. :)

**Comment by SteveJobzniak** [comment]
Thanks. Yeah I need to take it easy for a while haha. :)

**Comment by beshoo1** [comment]
@SteveJobzniak, i tried to be polite as much as possible. 
I am not a criminal, nor spamer. 
I may dump in your judgment, but for sure i am learning. 

If this place for pro. Then please BLOCK my user for good and let it say hi to the floor. 

Why i am using proxy, easy, service to offer statistics info for users, instagram will block it. Bot it may like a bot but not for   spam  purposes. 

 Again, I am not a spamer, and your words is very offensive. 
(You own me an apology.) 

Regarding if i can do this script or not, your scripting skills way better than me. 

Will i am speechless. 
Here in Syria we are in a middle of ware tring to survive, and struggleing to learn by ask who know.... 

But it seems even who know want to be an enemy too agents dump learners. 

Thank you for support.

**Comment by SteveJobzniak** [comment]
> Why i am using proxy, easy, service to offer statistics info for users, instagram will block it. Bot it may like a bot but not for spam purposes.

Wtf? Either you are a liar or stupid.

If you want to provide a legal service to offer statistics info for users, you simply have to sign up for OFFICIAL API ACCESS at Instagram:

https://www.instagram.com/developer/

For FREE.

I doubt that you're telling the truth. You spent 4 weeks trying to make Instagram-API hide your IP. And if you really are making a legal service, then you must sign up to their official API (which protects user logins/passwords by letting them authenticate to Instagram instead of you seeing their password).

Instead you are using this private, hacked API project, and doing your best to hide your IP. Whatever you're doing, I don't think it's legal. You might be planning to steal logins by having them "get stats" via your site, hence using our API instead of the official one (which won't let you steal their logins). Or you might be planning spam/other criminal activity.

Bye bye, you're getting banned now. Your story is too implausible to be anything legitimate.

**Comment by beshoo1** [comment]
Please kindly do not judge people and you owe me an apology. 

Any way thank you for your time.

**Comment by beshoo1** [comment]
Regarding official API provided by Instagram actually I did try to sign up but unfortunately Instagram did not accept because they want me to design an intro of the application. 

Which I am not the good in it
 and do you know Instagram set a limit for the request to Instagram server. 

Anyway I don't have to justify myself again

**Comment by romalytvynenko** [comment]
Hello guys!

Photos publishing to Instagram working well if photo size in Instagram aspects ratio bounds: from 5:4 (vertical photo) to 1.91:1 (horizontal).
